backend/routes/programs.js

```diff
@@ -39,3 +44,187 @@ router.post("/", function(req, res) {
 });
 
 module.exports = router;
+
+// Get all the users of a program
+// Returns an array of users: [{userId, email, name, role}]
+router.get("/:programId/users", function(req, res) {
+  db["Authentication"].users
+    .aggregate([
+      {
+        $match: {
+          programs: {
+            $elemMatch: {
+              id: req.params.programId
+            }
+          }
+        }
+      },
+      {
+        $unwind: {
+          path: "$programs"
+        }
+      },
+      {
+        $match: {
+          "programs.id": req.params.programId
+        }
+      },
+      {
+        $project: {
+          _id: 0,
+          userId: 1,
+          email: 1,
+          name: 1,
+          role: "$programs.role"
+        }
+      }
+    ])
+    .then(function(found) {
+      res.json(found);
+    })
+    .catch(function(err) {
+      res.send(err);
+    });
+});
+
+// Add a user to the program
+router.post("/:programId/users", async function(req, res) {
```

Comment:
Couple things

1. If the user is already assigned a role for this program, calling this will add another entry to their programs array. It should update their role if it already exists? OR throw an error since we already have an update method (unless you had a specific use case in mind?)



2. I am able to enter a role that doesn't exist in the `enum`. Normally mongo will throw an error when you try to do that and I am not sure why it didn't. 

3. I think it should be `/:programId/user` instead of `/:programId/users`. Just to make it clear that it is one user and not an array of users.

Reply:
1. Yeah, this one's pretty tricky. Ideally we have a validation error if this gets called and the user's already in the program (the UI will also prevent this). I've been playing around with some different options (Mongoose validators, `$addToSet`, etc.) but they don't work given this is an array of objects. I'll keep digging into it tomorrow!
2. Forgot to add `runValidators: true`! This should work now.
3. Will update this :)

Reply:
1. Mongo creates an `_id` for each sub-object. I am 88% sure that `$addToSet` checks the object exactly, including the `_id` which for obvious reasons will always be unique. This could be why but I don't know.

I think if you add `{_id: false}` to the schema it will disable the unique ids. I've never done this and not sure if it is a good practice or if it will solve anything.

2. Works!

3. Thanks!

Reply:
Yeah, I'm fairly certain that's why. Since we're tight on time, an easy solution could be to query the user's programs and make sure the program ID isn't already there. An extra query isn't ideal, but shouldn't be _too_ inefficient since users are only part of a few programs. If there's time, I can look into optimizing this later.

What do you think?

Reply:
Just pushed an easy fix ([commit](https://github.com/uwblueprint/cookie-monster/pull/293/commits/2481d1534045dbee1c4093e9a81b66ce13837128)), we can just let the upsert happen and handle the DuplicateKey error. Let me know what you think :)